ros2/ament.bzl

```diff
@@ -390,3 +391,89 @@ def sh_launcher(name, deps, idl_deps = None, **kwargs):
         ament_setup = ament_setup,
         **kwargs
     )
+
+def _ros2_ament_cpp_library_impl(ctx):
+    src_output = ctx.actions.declare_file(ctx.attr.basename + ".cpp")
+    hdr_output = ctx.actions.declare_file(ctx.attr.basename + ".hpp")
+    ament_prefix_path = ctx.attr.ament_setup[Ros2AmentSetupInfo].ament_prefix_path or ""
+
+    substitutions = dicts.add(
+        ctx.attr.substitutions,
+        {
+            "{{ament_prefix_path}}": ament_prefix_path,
+            "{{header}}": hdr_output.short_path,
+        },
+    )
+
+    expand_template_impl(
+        ctx,
+        template = ctx.file.src_template,
+        output = src_output,
+        substitutions = substitutions,
+        is_executable = False,
+    )
+    expand_template_impl(
+        ctx,
+        template = ctx.file.hdr_template,
+        output = hdr_output,
+        substitutions = substitutions,
+        is_executable = False,
+    )
+
+    files = depset([src_output, hdr_output])
+    return [
+        DefaultInfo(
+            files = files,
+        ),
+    ]
+
+ros2_ament_cpp_library_src = rule(
+    attrs = {
+        "ament_setup": attr.label(
+            mandatory = True,
+            providers = [Ros2AmentSetupInfo],
+        ),
+        "basename": attr.string(mandatory = True),
+        "data": attr.label_list(allow_files = True),
+        "hdr_template": attr.label(
+            mandatory = True,
+            allow_single_file = True,
+        ),
+        "src_template": attr.label(
+            mandatory = True,
+            allow_single_file = True,
+        ),
+        "substitutions": attr.string_dict(mandatory = True),
+    },
+    implementation = _ros2_ament_cpp_library_impl,
+)
+
+def ros2_ament_setup_library(name, deps, idl_deps = None, **kwargs):
+    ament_setup = name + "_ament_setup"
+    testonly = kwargs.get("testonly", False)
+    ros2_ament_setup(
+        name = ament_setup,
+        deps = deps,
+        idl_deps = idl_deps,
+        package_name = native.package_name(),
+        tags = ["manual"],
+        testonly = testonly,
+    )
+    cpp_namespace = native.package_name().replace("/", "_") + "_" + name
+    ros2_ament_cpp_library_src(
+        name = name + "_srcs",
```

Comment:
name + "_cpp_ament_setup" would be more consistent IMO.

Reply:
This target generates source files for the target with name `name`. I think that makes `_srcs` appropriate?

Reply:
OK, let's leave it as-is :+1: .